packages/flutter/lib/src/semantics/semantics.dart

```diff
@@ -1293,18 +1293,62 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
   }
 }
 
-/// A helper class to contain a semantics node and the effective
-/// sort order of that node during the traversal of the tree. This is
-/// what is actually sorted when semantics nodes are sorted.
+/// This class defines the comparison that is used to sort [SemanticsNode]s
+/// before sending them to the platform side.
+///
+/// It is a helper class used to contain a [SemanticsNode], the effective
+/// [SemanticsSortOrder], the globally transformed starting corner, and the
+/// containing node's [TextDirection] during the traversal of the semantics
+/// node tree.
 class _TraversalSortNode implements Comparable<_TraversalSortNode> {
-  _TraversalSortNode(this.node, this.order);
+  _TraversalSortNode(this.node, this.order, this.containerTextDirection, Matrix4 transform) {
+    globalStartCorner = _transformPoint(
+      containerTextDirection == TextDirection.ltr ? node.rect.topLeft : node.rect.topRight,
+      transform,
+    );
+  }
+
+  /// The node that this sort node represents.
   SemanticsNode node;
+
+  /// The effective text direction for this node is the directionality that
+  /// its container has.
+  TextDirection containerTextDirection;
+
+  /// This is the effective sort order for this node, taking into account its
+  /// parents.
   SemanticsSortOrder order;
 
+  /// The is the starting corner for the rectangle on this semantics node in
+  /// global coordinates. When the container has the directionality
+  /// [TextDirection.ltr], this is the upper left corner.  When the container
+  /// has the directionality [TextDirection.rtl], this is the upper right
+  /// corner.
+  Offset globalStartCorner;
+
+  Offset _transformPoint(Offset point, Matrix4 matrix) {
+    final Vector3 result = matrix.transform3(new Vector3(point.dx, point.dy, 0.0));
+    return new Offset(result.x, result.y);
+  }
+
+  /// Compares the node's start corner with that of `other`.
+  ///
+  /// Sorts top to bottom, and then start to end.
+  ///
+  /// Takes into account the container text direction, since the
+  /// coordinate system has zero on the left, and we need to compare
+  /// differently for different text directions.
   int _compareGeometry(_TraversalSortNode other) {
-    // TODO(gspencer): Move the geometric comparison from the platform side to here.
-    // This involves calculating the globally-transformed quad for the semantics node rect
-    // and then sorting by its bounding box, based on the container's directionality.
+    final int verticalDiff = globalStartCorner.dy.compareTo(other.globalStartCorner.dy);
+    if (verticalDiff != 0) {
+      return verticalDiff;
+    }
+    switch (containerTextDirection) {
+      case TextDirection.ltr:
+        return globalStartCorner.dx.compareTo(other.globalStartCorner.dx);
+      case TextDirection.rtl:
+        return other.globalStartCorner.dx.compareTo(globalStartCorner.dx);
```

Comment:
trivial nit: generally we prefer rtl to be first

Reply:
See discussion at https://master-docs-flutter-io.firebaseapp.com/flutter/dart-ui/TextDirection-class.html